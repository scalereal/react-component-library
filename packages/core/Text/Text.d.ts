import { FC } from 'react';
import { TextProps } from './types';
import { WithStyle } from '../../utils';
declare const Text: FC<TextProps> & WithStyle;
export default Text;
