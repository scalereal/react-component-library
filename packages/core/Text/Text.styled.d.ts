import { TextProps } from './types';
export declare const Text: import("styled-components").StyledComponent<"span", any, TextProps, never>;
