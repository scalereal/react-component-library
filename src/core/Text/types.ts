import { HTMLProps, WithThemeProp } from '../../utils';
import { TextSizes } from "../../theme/core/text/types"
import { TextWeights, TextAlignments } from '../../theme/core/text/types';
export interface TextProps extends HTMLProps<HTMLElement>, WithThemeProp {
    /**Text Size */
    textSize: TextSizes;
    /** Text color */
    textColor?: string;
    /** Text Uppercase*/
    uppercase?: boolean;
    /** Line through text */
    lineThrough?: boolean;
    /**Text Line Height */
    lineHeight?:string;
    /** Text Hover Color */
    hoverColor?:string;
    /** Text Active Color*/
    activeColor?:string;
    /** Text weight */
    textWeight?: TextWeights;
    /** Text alignment */
    textAlign?:TextAlignments;
}
