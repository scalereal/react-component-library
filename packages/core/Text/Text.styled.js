var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import styled, { css } from 'styled-components';
import { defaultTheme } from '../../theme';
var uppercase = function () { return css(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n    text-transform: uppercase;\n"], ["\n    text-transform: uppercase;\n"]))); };
var lineThrough = function () { return css(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n    text-decoration: line-through;\n"], ["\n    text-decoration: line-through;\n"]))); };
var getTextStyle = function (_a) {
    var props = __rest(_a, []);
    var textColor = props.textColor, lineHeight = props.lineHeight, textSize = props.textSize, textAlign = props.textAlign, textWeight = props.textWeight;
    return css(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n        margin: 0;\n        color: ", ";\n        line-height: ", ";\n        font-size: ", ";\n        text-align: ", ";\n        font-weight: ", ";\n        font-family: ", ";\n    "], ["\n        margin: 0;\n        color: ", ";\n        line-height: ", ";\n        font-size: ", ";\n        text-align: ", ";\n        font-weight: ", ";\n        font-family: ", ";\n    "])), function (_a) {
        var theme = _a.theme;
        return (textColor ? textColor : theme.text.defaults.textColor);
    }, function (_a) {
        var theme = _a.theme;
        return (lineHeight ? lineHeight : theme.text.defaults.lineHeight);
    }, function (_a) {
        var theme = _a.theme;
        return (textSize === undefined ? theme.text.defaults.size : theme.text.sizes[textSize].fontSize);
    }, function (_a) {
        var theme = _a.theme;
        return (textAlign ? textAlign : theme.text.defaults.textAlign);
    }, function (_a) {
        var theme = _a.theme;
        return (textWeight ? theme.text.weights[textWeight].weight : theme.text.defaults.textWeight);
    }, function (_a) {
        var theme = _a.theme;
        return (theme.text.fontFamily);
    });
};
export var Text = styled('span')(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n    padding:0.3rem;\n    ", "\n    ", ";\n    ", ";\n    &:active {\n        color: ", ";\n    }\n    ", ";\n    /* &:hover {\n        color: ", ";\n    } */\n\n"], ["\n    padding:0.3rem;\n    ", "\n    ", ";\n    ", ";\n    &:active {\n        color: ", ";\n    }\n    ",
    ";\n    /* &:hover {\n        color: ", ";\n    } */\n\n"])), getTextStyle, function (props) { return props.uppercase && uppercase(); }, function (props) { return props.lineThrough && lineThrough(); }, function (_a) {
    var theme = _a.theme, activeColor = _a.activeColor;
    return (activeColor ? activeColor : theme.text.defaults.activeColor);
}, function (props) { return props.hoverEffect ? "&:hover {\n        color: " + (props.hoverColor ? props.hoverColor : props.theme.text.defaults.hoverColor) + ";\n    }" : ''; }, function (_a) {
    var theme = _a.theme, hoverColor = _a.hoverColor;
    return (hoverColor ? hoverColor : theme.text.defaults.hoverColor);
});
Text.displayName = "Text";
Text.defaultProps = {
    theme: defaultTheme
};
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
