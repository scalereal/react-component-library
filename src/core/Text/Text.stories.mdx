import { Text, textSizes, textAlignments, textWeights } from './Text';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Text"
    component={Text}
    argTypes={{
        hoverEffect: {
            control: { type: 'boolean' }
        },
        textColor: {
            control: 'color'
        },
        hoverColor: {
            control: 'color'
        },
        activeColor: {
            control: 'color'
        },
        textSize: {
            control: {
                type: 'select',
                options: Object.values(textSizes)
            }
        },
        textAlign: {
            control: {
                type: 'select',
                options: Object.values(textAlignments)
            }
        },
        textWeight: {
            control: {
                type: 'select',
                options: Object.values(textWeights)
            }
        }
    }}
/>

# Text

Displays a Text

export const Template = ({
    textColor,
    activeColor,
    hoverColor,
    lineThrough,
    uppercase,
    textSize,
    lineHeight,
    textWeight,
    textAlign,
    ...args
}) => {
    return (
        <Text
            textColor={textColor}
            activeColor={activeColor}
            hoverColor={hoverColor}
            lineThrough={lineThrough}
            uppercase={uppercase}
            textSize={textSize}
            lineHeight={lineHeight}
            textWeight={textWeight}
            textAlign={textAlign}
            {...args}
        >
            Lorem Ipsum is simply dummy text
        </Text>
    );
};

<Preview>
    <Story
        name="Text"
        args={{
            textSize: 'S'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Text textColor="blue">Lorem Ipsum is simply dummy text</Text>
        <Text textColor="green">Lorem Ipsum is simply dummy text</Text>
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Text textSize="XS">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="S">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="M">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="L">Lorem Ipsum is simply dummy text</Text>
        <Text textSize="XL">Lorem Ipsum is simply dummy text</Text>
    </div>
</Preview>

### Props

<Props story="Text" />
