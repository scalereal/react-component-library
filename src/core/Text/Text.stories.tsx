import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Text from "./Text"


export default {
    title: 'Components/Core/Text',
    component: Text,
    argTypes : {
        textColor: {
            control:'color'
        },
    }
} as Meta

const Template:Story = ({ textColor, lineThrough, uppercase, textSize, lineHeight, ...args}) => {
    return <Text textColor={textColor} lineThrough={lineThrough} uppercase={uppercase} textSize={textSize} lineHeight={lineHeight} {...args}>
    Lorem Ipsum is simply dummy text
    </Text>
}

export const Default = Template.bind({});
Default.args = {
    textColor:"#435465",
    textSize:'S'
}
