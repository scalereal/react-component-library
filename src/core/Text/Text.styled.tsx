import styled, { css } from 'styled-components';
import { defaultTheme } from '../../theme';
import { TextProps } from './types';

const uppercase = () => css`
    text-transform: uppercase;
`;
const lineThrough = () => css`
    text-decoration: line-through;
`;

const getTextStyle = ({ ...props }: TextProps) => {
    const { textColor, lineHeight, textSize, textAlign, textWeight } = props;
    return css`
        margin: 0;
        color: ${({ theme }) => (textColor ? textColor : theme.text.defaults.textColor)};
        line-height: ${({ theme }) => (lineHeight ? lineHeight : theme.text.defaults.lineHeight)};
        font-size: ${({ theme }) => (textSize === undefined ? theme.text.defaults.size :theme.text.sizes[textSize].fontSize)};
        text-align: ${({ theme }) => (textAlign ? textAlign : theme.text.defaults.textAlign)};
        font-weight: ${({ theme })=> (textWeight ? theme.text.weights[textWeight].weight : theme.text.defaults.textWeight)};
        font-family: ${({ theme }) => (theme.text.fontFamily)};
    `;
};

export const Text = styled('span')<TextProps>`
    padding:0.3rem;
    ${getTextStyle}
    ${props => props.uppercase && uppercase()};
    ${props => props.lineThrough && lineThrough()};
    &:active {
        color: ${({ theme, activeColor }) => (activeColor ? activeColor : theme.text.defaults.activeColor)};
    }
    ${props => props.hoverEffect? `&:hover {
        color: ${props.hoverColor ? props.hoverColor : props.theme.text.defaults.hoverColor};
    }`:''
    };
    /* &:hover {
        color: ${({ theme, hoverColor }) => (hoverColor ? hoverColor : theme.text.defaults.hoverColor)};
    } */
`;

Text.displayName = "Text";
Text.defaultProps = {
    theme:defaultTheme
}
