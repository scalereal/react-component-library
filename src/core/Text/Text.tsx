import React, { FC } from 'react'
import { WithStyle } from '../../utils'
import * as Styled from './Text.styled'
import { TextProps } from './types'

export const textSizes : TextProps['textSize'][] = ['XS', 'S', 'M', 'L', 'XL']
export const textAlignments : TextProps['textAlign'][] = ['left', 'right', 'center', 'justify', 'justify-all', 'start', 'end', 'match-parent', 'inherit', 'initial', 'unset']
export const textWeights : TextProps['textWeight'][] = ['light', 'regular', 'medium', 'strong', 'extraStrong']

const Text: FC<TextProps> & WithStyle = React.memo(
    React.forwardRef(({ children, textSize, textWeight, textAlign, ...restProps }, ref) => {
        return (
            <Styled.Text  textSize={textSize} textWeight={textWeight} textAlign={textAlign} ref={ref} {...restProps }>
                {children}
            </Styled.Text>
        );
    })
);

Text.defaultProps = {
    lineThrough: false,
    uppercase: false,
    textSize:'S'
};
Text.displayName = 'Text';
Text.Style = Styled.Text;

export default Text;
