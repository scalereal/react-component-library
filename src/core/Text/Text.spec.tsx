import { render } from '@testing-library/react'
import React from 'react'
import Text  from './Text'

describe('Text', () => {
    it('should render span element by default', () => {
        const { container } = render(<Text textSize="S" />);
        expect(container).toMatchSnapshot();
    });

    it('should render with all the props', () => {
        const { container } = render(<Text activeColor="red" hoverColor="blue" textColor="pink" uppercase lineThrough textSize="S" lineHeight="1" />);
        expect(container).toMatchSnapshot();
    });
});
