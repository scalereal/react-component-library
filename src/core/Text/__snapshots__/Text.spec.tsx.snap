// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should render span element by default 1`] = `
<div>
  <span
    class="sc-bdfBwQ hcPyHd"
  />
</div>
`;

exports[`Text should render with all the props 1`] = `
<div>
  <span
    class="sc-bdfBwQ bttaXt"
  />
</div>
`;
